### PR TITLE
Cgroup2: Reduce allocations for `manager.Stat`

### PR DESCRIPTION
**Pull request by dcantah**
Related to https://github.com/containerd/cgroups/issues/280. There's still some spots we can fix, as the majority of allocs are occurring in bufio.Scan in readKVStatsFile. cc @kolyshkin @manugupt1 seeming as how we were all there for the last one 

This change works towards bringing down allocations for manager.Stat(). There's quite a few things we can do:

1. Swap to opening a file and issuing a single read for uint64/"max" values. Previously we were doing os.ReadFile's which returns a byte slice, so it needs to allocate.
2. Sizing the map we store {controller}.stat values in. We store 40+ stats in this map and the default bucket size for Go seems to be smaller than this, so we'd incur a couple allocs at runtime when adding these.
3. Finally, and funnily enough the biggest win, was just not calling f.Readdir(-1) for hugeTlb stats, and swapping this for a trick the runc cgroup code does. The number of allocs shaved off here is tied to the number of files in the cgroup directory
as f.Readdir(-1) was being called everytime we went to collect HugeTlb stats (it returns a slice of interfaces, that all we were doing is using the Name() method of anyways). To optimize this we can take advantage of the fact that we know exactly what files we need to open as they're fixed, the only variable portion is the hugepage sizes available on the host. To get around this we can compute the hugepage sizes in a sync.Once, and all future manager.Stat calls can reap the benefits as we don't need to Readdir() the entire cgroup path to search for hugetlb.pagesize.foobar's.

All in all on Go 1.19.4 this drops allocations in Stat from 322 to 198

```shell
benchstat ~/benchstat/old.txt ~/benchstat/new.txt 
goos: linux
goarch: amd64
pkg: github.com/containerd/cgroups/v3/cgroup2
cpu: AMD Ryzen 9 5900X 12-Core Processor            
        │ /home/dcantah/benchstat/old.txt │   /home/dcantah/benchstat/new.txt   │
        │             sec/op              │   sec/op     vs base                │
Stat-24                      112.29µ ± 3%   65.10µ ± 6%  -42.02% (p=0.000 n=10)

        │ /home/dcantah/benchstat/old.txt │   /home/dcantah/benchstat/new.txt    │
        │              B/op               │     B/op      vs base                │
Stat-24                      41.40Ki ± 0%   22.26Ki ± 0%  -46.23% (p=0.000 n=10)

        │ /home/dcantah/benchstat/old.txt │  /home/dcantah/benchstat/new.txt   │
        │            allocs/op            │ allocs/op   vs base                │
Stat-24                        322.0 ± 0%   198.0 ± 0%  -38.51% (p=0.000 n=10)
```

### CODE REVIEW

**Comment by dcantah** [comment]
Working towards https://github.com/containerd/cgroups/issues/280. There's still some spots we can fix, as the majority of allocs are occurring in bufio.Scan in readKVStatsFile

**Comment by dcantah** [comment]
cc @kolyshkin @manugupt1 seeming as how we were all there for the last one 🤣

File: cgroup2/manager.go [hunk 1]
```diff
@@ -763,37 +755,42 @@ func (c *Manager) EventChan() (<-chan Event, <-chan error) {
...
 
-func parseMemoryEvents(out map[string]interface{}) (Event, error) {
+func parseMemoryEvents(out map[string]string) (Event, error) {
 	e := Event{}
 	if v, ok := out["high"]; ok {
-		e.High, ok = v.(uint64)
-		if !ok {
+		val, err := parseUint(v, 10, 64)
+		if err != nil {
 			return Event{}, fmt.Errorf("cannot convert high to uint64: %+v", v)
```
Comment (kolyshkin) [inline cgroup2/manager.go:763]: Not related to this PR, but it would be easier to return `nil`not `Event{}` every time an error is returned

File: cgroup2/utils_test.go [hunk 2]
```diff
@@ -64,3 +64,22 @@ func TestToResources(t *testing.T) {
...
+		_ = getStatFileContentUint64("/proc/self/loginuid")
+	}
+}
+
+func BenchmarkReadSingleFile(b *testing.B) {
+	b.ReportAllocs()
+
+	out := make(map[string]string)
+	for i := 0; i < b.N; i++ {
+		if err := readSingleFile("/proc/self/loginuid", "", out); err != nil {
```
Comment (kolyshkin) [inline cgroup2/utils_test.go:81]: readSingleFile expects a file with key value pairs, so `/proc/self/loginuid` is not a good choice.

File: cgroup2/utils.go [hunk 3]
```diff
@@ -242,21 +237,30 @@ func ToResources(spec *specs.LinuxResources) *Resources {
...
 		return math.MaxUint64
 	}
 
 	res, err := parseUint(trimmed, 10, 64)
 	if err != nil {
-		logrus.Errorf("unable to parse %q as a uint from Cgroup file %q", string(contents), filePath)
+		logrus.Errorf("unable to parse %q as a uint from Cgroup file %q", trimmed, filePath)
 		return res
 	}
-
```
Comment (kolyshkin) [inline cgroup2/utils.go:259]: this should not be here

**Comment by kolyshkin** [comment]
I suggest making `readHugeTlbStats` changes related to readdirnames as a separate commit.

**Comment by dcantah** [comment]
> I suggest making `readHugeTlbStats` changes related to readdirnames as a separate commit.

That's fair, they were all in the pursuit of improving Stat so I felt either way. I originally had the readSingleFile and uint improvements in a separate commit and squished after finding the other areas. I'll separate them though

File: cgroup2/utils_test.go (hunk 2 above)
Comment (dcantah) [reply cgroup2/utils_test.go:81]: It doesn't afaict, it just places the resulting value in the passed in map. The original logic was to try and parse the output as a uint, and if it failed just treat the output as a string. It's only called for pids.current and pids.max which are both either "max" or some unsigned integer

File: cgroup2/manager.go (hunk 1 above)
Comment (dcantah) [reply cgroup2/manager.go:763]: Agree, I'm going to leave though to avoid too much refactor off the main goal. We can always do a followup if need be

File: cgroup2/utils.go (hunk 3 above)
Comment (dcantah) [reply cgroup2/utils.go:259]: Ack, fixing in a little!

**Comment by dcantah** [comment]
@kolyshkin Made a separate commit here for the hugeTlb changes, thanks for the quick review!

File: cgroup2/manager.go [hunk 4]
```diff
@@ -529,7 +528,10 @@ func (c *Manager) Stat() (*stats.Metrics, error) {
 	if err != nil {
 		return nil, err
 	}
-	out := make(map[string]interface{})
+	// Sizing this avoids an allocation to increase the map at runtime;
+	// currently the default bucket size is 8 and we put 40+ elements
+	// in it so we'd always end up allocating.
+	out := make(map[string]string, 50)
```
Comment (kolyshkin) [inline cgroup2/manager.go:534]: perhaps have the size as 0 and capacity as 64?

File: cgroup2/utils_test.go (hunk 2 above)
Comment (kolyshkin) [reply cgroup2/utils_test.go:81]: Sorry, I mixed that up with readKVStatsFile. Yes, this is the right approach then.

File: cgroup2/manager.go (hunk 4 above)
Comment (dcantah) [reply cgroup2/manager.go:534]: You can't set the capacity of a map or chan, only slices

**Comment by kolyshkin** [comment]
By looking at the code, I'd rather have it refactored first, then optimize. At least some parts that are crying to be removed, such as readSingleFile -- PTAL https://github.com/containerd/cgroups/pull/283

**Comment by dcantah** [comment]
@kolyshkin ack, there is quite a few places that all do the same thing. I'm fine getting rid of them and then rebasing here

File: cgroup2/manager.go (hunk 4 above)
Comment (kolyshkin) [reply cgroup2/manager.go:534]: Right, my bad, ENEEDCOFFEE

File: cgroup2/manager.go [hunk 5]
```diff
@@ -625,56 +627,53 @@ func (c *Manager) Stat() (*stats.Metrics, error) {
...
 	f, err := os.Open(filepath.Join(path, file))
 	if err != nil {
 		return err
 	}
 	defer f.Close()
-	data, err := io.ReadAll(f)
+
+	// We expect an unsigned 64 bit integer, or a "max" string
+	// in some cases.
+	buf := make([]byte, 32)
```
Comment (manugupt1) [inline cgroup2/manager.go:666]: Is this way too low or
https://github.com/containerd/cgroups/pull/278/files#diff-f6fc333a9f621c02b1a9723d8cae6ad3ddd6cdc2501c75cc2a5765bbf01e631cR140 this too high?

File: cgroup2/utils_test.go [hunk 6]
```diff
@@ -64,3 +64,22 @@ func TestToResources(t *testing.T) {
 	v2resources2 := ToResources(&res2)
 	assert.Equal(t, CPUMax("max 10000"), v2resources2.CPU.Max)
 }
+
+func BenchmarkReaduint64(b *testing.B) {
```
Comment (manugupt1) [inline cgroup2/utils_test.go:68]: nit: BenahcMarkGetStatFileContentUint64

File: cgroup2/manager.go (hunk 5 above)
Comment (dcantah) [reply cgroup2/manager.go:666]: Yea, I think it's too high. It was originally a brainfart on my part of attributing 64 bit with needing 64 bytes for the character representation in the file, and I found out what the trimming of space was likely for for all of these parsing functions. All of the cgroup files seem to end with a single line feed, and TrimSpace counts this as white space, so it's basically just removing \n for any of the files.  We should only need a buffer of 20 to handle the max uint64, so 32's just for power of 2. I'll open a change to lower the other buffer

**Comment by dcantah** [comment]
Drafting this till https://github.com/containerd/cgroups/pull/283 gets in

File: cgroup2/utils.go [hunk 7]
```diff
@@ -379,41 +389,55 @@ func systemdUnitFromPath(path string) string {
...
 	usage := []*stats.HugeTlbStat{}
 	keyUsage := make(map[string]*stats.HugeTlbStat)
+
 	f, err := os.Open(path)
 	if err != nil {
 		return usage
 	}
-	files, err := f.Readdir(-1)
+
+	files, err := f.Readdirnames(-1)
```
Comment (manugupt1) [inline cgroup2/utils.go:398]: You can change to improve further marginally by using f.ReadDir(-1).
See: https://go.dev/play/p/aMmQKCKBZpN

I got 
```
-timeout 30m

goos: linux
goarch: amd64
pkg: github.com/manugupt1/file
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkReaddir-8        	     871	   1300868 ns/op	  130113 B/op	    1439 allocs/op
BenchmarkReadDir-8        	    3076	    370488 ns/op	   48972 B/op	     957 allocs/op
BenchmarkReadDirNames-8   	    3139	    366333 ns/op	   48973 B/op	     957 allocs/op
PASS
	github.com/manugupt1/file	coverage: 50.0% of statements
ok  	github.com/manugupt1/file	3.655s
```

File: cgroup2/utils.go [hunk 8]
```diff
@@ -379,41 +389,55 @@ func systemdUnitFromPath(path string) string {
...
 				hugeTlb = &stats.HugeTlbStat{}
 			}
 			hugeTlb.Pagesize = pageSize
-			out, err := os.ReadFile(filepath.Join(path, file.Name()))
+
+			f, err := os.Open(filepath.Join(path, file))
 			if err != nil {
 				continue
 			}
+			defer f.Close()
```
Comment (manugupt1) [inline cgroup2/utils.go:423]: nit: Can/Should we instead try to close the file pointer in the function after reading rather than defering the call?

File: cgroup2/utils.go (hunk 7 above)
Comment (dcantah) [reply cgroup2/utils.go:398]: Hmm, this is what I swapped to originally but found the latter shaved down a few more allocations. Let me try again

Comment (manugupt1) [reply cgroup2/utils.go:398]: Hmm! I am not sure why; maybe a typo : ReadDir vs Readdir, because ReadDir is further down the stack when you look through go's source code.

File: cgroup2/utils.go [hunk 9]
```diff
@@ -379,41 +389,55 @@ func systemdUnitFromPath(path string) string {
...
+
+			n, err := f.Read(buf)
+			if err != nil {
+				continue
+			}
+
 			var value uint64
-			stringVal := strings.TrimSpace(string(out))
+			stringVal := strings.TrimSpace(string(buf[:n]))
```
Comment (manugupt1) [inline cgroup2/utils.go:431]: Another realization: if we use the bytes package; its again faster; does not change allocs though.
See: https://go.dev/play/p/7jPtYNfN7xq

I get 
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^(BenchmarkByteCompare|BenchmarkStringCompare)$ github.com/manugupt1/file -count=1 -timeout 30m

goos: linux
goarch: amd64
pkg: github.com/manugupt1/file
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkByteCompare-8     	64981508	        17.42 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringCompare-8   	41275358	        27.36 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/manugupt1/file	3.223s
```

Comment (manugupt1) [reply cgroup2/utils.go:431]: I will do these for cgroup1 if you are fine with it.

File: cgroup2/utils.go (hunk 7 above)
Comment (dcantah) [reply cgroup2/utils.go:398]: No it was definitely ReadDir that I tried as I noticed the stdlib comment on Readdir originally which sparked this investigation https://pkg.go.dev/os#File.Readdir:

"Most clients are better served by the more efficient ReadDir method." What Go version are you on? Should upgrade to 1.20 on my end on my linux machine 😳

Comment (manugupt1) [reply cgroup2/utils.go:398]: Ah I am on go1.20 that's it

File: cgroup2/utils.go (hunk 9 above)
Comment (dcantah) [reply cgroup2/utils.go:431]: Go for it! That's a good point, we can defer the converting to strings after we trim

File: cgroup2/utils.go (hunk 8 above)
Comment (dcantah) [reply cgroup2/utils.go:423]: You mean manual f.Close calls if Read fails/if read succeeds?

File: cgroup2/utils.go (hunk 7 above)
Comment (kolyshkin) [reply cgroup2/utils.go:398]: I'm thinking maybe we should use an alternative approach here. The thing is, all the file names are well known ahead of time. The only thing that needs to be known is the list of hugetlb sizes (and for that, there's a function in cgroup1 package available (which needs to be wrapped in `once.Do`, but that's a different topic)).

When we know all page sizes, the code becomes something like this (an example from runc/libcontainer):

https://github.com/opencontainers/runc/blob/3a2c0c2565644d8a7e0f1dd594a060b21fa96cf1/libcontainer/cgroups/fs/hugetlb.go#L31-L59

Comment (dcantah) [reply cgroup2/utils.go:398]: That does seem more ideal. I'll play with this tonight, thanks!

Comment (kolyshkin) [reply cgroup2/utils.go:398]: As for the three versions of reading the file names,
- `os.Readdir` is the worst, because it calls `stat(2)` on every file;
- `os.ReadDir` is better as it doesn't do `stat` but instead wraps each entry into FileInfo interface;
- `os.Readdirnames` is the fastest as it just returns the list of names.

Since the name is everything we care about here, `os.Readdirnames` is the best.

Yet I would rewrite this using the approach without reading the directly (see above).

File: cgroup2/utils.go [hunk 10]
```diff
@@ -389,41 +389,55 @@ func systemdUnitFromPath(path string) string {
...
 	}
 
+	buf := make([]byte, 32)
 	for _, file := range files {
-		if strings.Contains(file.Name(), "hugetlb") &&
-			(strings.HasSuffix(file.Name(), "max") || strings.HasSuffix(file.Name(), "current")) {
-			var hugeTlb *stats.HugeTlbStat
-			var ok bool
-			fileName := strings.Split(file.Name(), ".")
+		if strings.Contains(file, "hugetlb") &&
```
Comment (kolyshkin) [inline cgroup2/utils.go:406]: While at it, makes sense to switch to `strings.HasPrefix`, and add a `.` after `hugetlb`.

File: cgroup2/utils.go [hunk 11]
```diff
@@ -389,41 +389,55 @@ func systemdUnitFromPath(path string) string {
...
+			(strings.HasSuffix(file, "max") || strings.HasSuffix(file, "current")) {
```
Comment (kolyshkin) [inline cgroup2/utils.go:407]: same here -- these better be `.max` and `.current`.

File: cgroup2/utils.go (hunk 10 above)
Comment (dcantah) [reply cgroup2/utils.go:406]: ack. I usually would leave things like this for another cleanup round but I'll just make another commit on top here. We're already screwing around in this area so it's worthwhile

File: cgroup2/utils.go [hunk 12]
```diff
@@ -389,41 +389,55 @@ func systemdUnitFromPath(path string) string {
...
+			fileName := strings.Split(file, ".")
 			pageSize := fileName[1]
 			if hugeTlb, ok = keyUsage[pageSize]; !ok {
 				hugeTlb = &stats.HugeTlbStat{}
 			}
 			hugeTlb.Pagesize = pageSize
-			out, err := os.ReadFile(filepath.Join(path, file.Name()))
+
+			f, err := os.Open(filepath.Join(path, file))
 			if err != nil {
```
Comment (kolyshkin) [inline cgroup2/utils.go:420]: Hmm, why not use `getStatFileContentUint64` from here?

Comment (dcantah) [reply cgroup2/utils.go:420]: No reason other than forgetting 😆

File: cgroup2/utils.go (hunk 8 above)
Comment (kolyshkin) [reply cgroup2/utils.go:423]: @manugupt1 is right, because otherwise we may have quite a few files opened before we exit from this function.

Yet better -- switch to `getStatFileContentUint64`

Comment (dcantah) [reply cgroup2/utils.go:423]: ^ That seems best

File: cgroup2/utils.go (hunk 7 above)
Comment (manugupt1) [reply cgroup2/utils.go:398]: I read wrong again :(

**Comment by dcantah** [comment]
Okay, updated to include the runc trick @kolyshkin, great suggestion. We're down to 199 now.

File: cgroup2/utils.go [hunk 13]
```diff
@@ -378,55 +390,92 @@ func systemdUnitFromPath(path string) string {
...
-	files, err := f.Readdir(-1)
-	f.Close()
-	if err != nil {
-		return usage
+	for _, pagesize := range hugePageSizes() {
+		usage = append(usage, &stats.HugeTlbStat{
+			Max:      getStatFileContentUint64(filepath.Join(path, "hugetlb."+pagesize+".max")),
+			Current:  getStatFileContentUint64(filepath.Join(path, "hugetlb."+pagesize+".current")),
+			Pagesize: pagesize,
+		})
```
[the remark this answers is not in the document]
Comment (manugupt1) [reply cgroup2/utils.go:398]: Hmm! Can we do soemthing like
```
hugePageSizeSlice := hugePageSizes()
var usage = make(*stats.HugeTlbStat{}, len(hugePageSizeSlice)
and then iterate like an array
for idx, pageSize  := range hugePageSizeSlice {
  usage[idx] = .... 
}
```

File: cgroup2/manager_test.go [hunk 14]
```diff
@@ -223,3 +223,22 @@ func TestCgroupType(t *testing.T) {
...
+
+func BenchmarkStat(b *testing.B) {
+	checkCgroupMode(b)
+	group := "/stat-test-cg"
+	groupPath := fmt.Sprintf("%s-%d", group, os.Getpid())
+	c, err := NewManager(defaultCgroup2Path, groupPath, &Resources{})
+	if err != nil {
+		b.Fatal("failed to init new cgroup manager: ", err)
+	}
+	defer os.Remove(c.path)
```
Comment (manugupt1) [inline cgroup2/manager_test.go:235]: nit: b.Cleanup

File: cgroup2/testutils_test.go [hunk 15]
```diff
@@ -27,14 +27,14 @@ import (
 	"golang.org/x/sys/unix"
 )
 
-func checkCgroupMode(t *testing.T) {
+func checkCgroupMode(tb testing.TB) {
 	var st unix.Statfs_t
-	err := unix.Statfs(defaultCgroup2Path, &st)
-	require.NoError(t, err, "cannot statfs cgroup root")
-
+	if err := unix.Statfs(defaultCgroup2Path, &st); err != nil {
```
[the remark this answers is not in the document]
Comment (dcantah) [reply cgroup2/testutils_test.go:32]: Oh that's weird, my editor was yelling at me originally when I tried putting `require.NoError(tb, err)` here, but trying on my personal pc works fine..? require takes in `require.TestingT`'s which is just an interface wrapper defined as:

```go
type TestingT interface {
	Errorf(format [string](https://pkg.go.dev/builtin#string), args ...interface{})
	FailNow()
}
```

and testing.TB has both defined so it should've always worked 🤷‍♂️. Changing

File: cgroup2/utils.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (dcantah) [reply cgroup2/utils.go:398]: This is the same outcome though right? We still have a slice with empty entries, where the number of them is tied to how many huge page sizes we found in `/sys/kernel/mm/hugepages`. Before this we'd return an empty slice which I think is ideal

File: cgroup2/utils.go [hunk 16]
```diff
@@ -378,55 +390,92 @@ func systemdUnitFromPath(path string) string {
...
+	var warn error
+
+	for _, file := range fileNames {
+		// example: hugepages-1048576kB
+		val := strings.TrimPrefix(file, "hugepages-")
+		if len(val) == len(file) {
+			// Unexpected file name: no prefix found, ignore it.
+			continue
+		}
+		// The suffix is always "kB" (as of Linux 5.13). If we find
```
[the remark this answers is not in the document]
Comment (dcantah) [reply cgroup2/utils.go:447]: The comment is meant to mean as of 5.13 there's nothing but kB, not that 5.13 changed this to be all kB

File: cgroup2/utils.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (manugupt1) [reply cgroup2/utils.go:398]: Oh okay! yes. what I suggested matches your current code; but I was looking from an allocation perspective. If you pre-allocate; the slice won't have to expand while looping and appending.

I do not have an opinion on functionality as I do not know how users use it. I would try and keep the previous behavior though.

[the remark this answers is not in the document]
Comment (kolyshkin) [reply cgroup2/utils.go:398]: I agree with @manugupt1 that pre-allocating the `usage` slice is better than growing it dynamically. Even if it won't shave off any allocations, this is more optimized overall, and since we know the size, it makes total sense to do so.

File: cgroup2/utils.go [hunk 17]
```diff
@@ -388,55 +390,92 @@ func systemdUnitFromPath(path string) string {
...
+		}
+		files, err := dir.Readdirnames(0)
+		dir.Close()
+		if err != nil {
+			return
+		}
+
+		hPageSizes, err = getHugePageSizeFromFilenames(files)
+		if err != nil {
+			logrus.Warnf("HugePageSizes: %s", err)
```
Comment (kolyshkin) [inline cgroup2/utils.go:429]: nit: `s/Huge/huge/`

File: cgroup2/utils.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (kolyshkin) [reply cgroup2/utils.go:447]: We can say something like "in all known Linux versions up to 6.3" (it's still always `kB` in there, I just checked).

**Review by kolyshkin — COMMENTED** [review]
LGTM overall, except for preallocating the usage slice would be better.

@dcantah also, if you like, you can add `Co-authored-by: Kir Kolyshkin <kolyshkin@gmail.com>` to the second commit, as I wrote that venerable /sys/kernel/mm/hugepages/* parsing code.

**Comment by kolyshkin** [comment]
> All in all on Go 1.19.4 this drops allocations in Stat from 322 to 199, or 47%.

Oh, and if you can use `benchstat` for such calculations that'd be terrific.

File: cgroup2/utils.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (dcantah) [reply cgroup2/utils.go:398]: Oh I agree about pre-allocating, more the point I was trying to get across still stands: this behavior is different than whats here today. We'll have entries in hugetlb stats regardless of if someone even has the controller enabled, as long as there's entries in /sys/kernel/mm/hugepages

**Comment by dcantah** [comment]
@kolyshkin Sure thing! I didn't even realize you wrote that, so you've already gone through this pain 🤣

File: cgroup2/utils.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (manugupt1) [reply cgroup2/utils.go:398]: Right! I am not the best judge of this tbh!

[the remark this answers is not in the document]
Comment (dcantah) [reply cgroup2/utils.go]: Lets see what others think, I don't think this is too crazy a departure

**Comment by dcantah** [comment]
> Oh, and if you can use `benchstat` for such calculations that'd be terrific.

No wonder you recommended, I fat fingered 47 from 38.. No more programming past 12 🐸

**Review by fuweid — APPROVED** [review]
LGTM

**Review by estesp — APPROVED** [review]
